File: 3181-find-building-where-alice-and-bob-can-meet/3181-find-building-where-alice-and-bob-can-meet.cpp
class Solution {
private:
    int search(int height, vector<pair<int, int>>& monoStack){
        int ans = -1, left = 0, right = monoStack.size() - 1;
        while (left <= right){
            int mid = left + ((right - left) >> 1);

            if (monoStack[mid].first > height){
                ans = max(ans, mid);
                left = mid + 1;
            }
            else
                right = mid - 1;
        }
        return ans;
    }

public:
    vector<int> leftmostBuildingQueries(vector<int>& heights, vector<vector<int>>& queries) {
        int m = heights.size(), n = queries.size();
        vector<vector<pair<int, int>>> newQueries(m);
        vector<int> ans(n, -1);

        for (int i = 0; i < n; ++i){
            int a = queries[i][0], b = queries[i][1];
            if (a > b) swap(a, b);
            if (a == b || heights[a] < heights[b])
                ans[i] = b;
            else
                newQueries[b].push_back({heights[a], i});
        }

        vector<pair<int, int>> monoStack;
        for (int i = m - 1; i >= 0; --i){
            int monoStackSize = monoStack.size();
            for (auto& [a, b] : newQueries[i]){
                int pos = search(a, monoStack);
                if (pos >= 0 && pos < monoStack.size())
                    ans[b] = monoStack[pos].second;
            }
            while (!monoStack.empty() && monoStack.back().first <= heights[i])
                monoStack.pop_back();

            monoStack.push_back({heights[i], i});
        }
        return ans;
    }
};

File: 3181-find-building-where-alice-and-bob-can-meet/3181-find-building-where-alice-and-bob-can-meet.py
class Solution:
    def leftmostBuildingQueries(self, heights: List[int], queries: List[List[int]]) -> List[int]:
        def search(height: int) -> int:
            ans, left, right = -1, 0, len(stack) - 1

            while left <= right:
                mid = (left + right) >> 1
                if stack[mid][0] > height:
                    ans = max(ans, mid)
                    left = mid + 1
                else:
                    right = mid - 1
            return ans

        m, n = len(heights), len(queries)
        new_queries, stack = [[] for _ in range(m)], []
        ans = [-1] * n
        
        for curr, (a, b) in enumerate(queries):
            if a > b:
                a, b = b, a

            if a == b or heights[a] < heights[b]:
                ans[curr] = b
            else:
                new_queries[b].append((heights[a], curr))
            
        for i in range(m - 1, -1, -1):
            stack_sz = len(stack)

            for a, b in new_queries[i]:
                pos = search(a)

                if pos < stack_sz and pos >= 0:
                    ans[b] = stack[pos][1]
            
            while stack and stack[-1][0] <= heights[i]:
                stack.pop()
            stack.append((heights[i], i))

        return ans